aten/src/ATen/native/TensorCompare.cpp

```diff
@@ -355,15 +377,30 @@ std::vector<Tensor> where(const Tensor& condition) {
 }
 
 Tensor _s_where(const Tensor& condition, const Tensor& self, const Tensor& other) {
-  TORCH_CHECK(self.dtype() == other.dtype(), "expected scalar type ", self.dtype(), " but found ", other.dtype());
-  Tensor ret = at::empty(self.sizes(), self.options());
+  // Here we do the type promotion without the TensorIterator machinery.
+  // For `where` we only want to consider `self` and `other`,
+  // however right now TensorIterator API doesn't have a way to specify which inputs
+  // should take part in type promotion.
+  auto common_dtype = at::result_type(self, other);
+  Tensor ret = at::empty(self.sizes(), self.options().dtype(common_dtype));
+
+  auto get_promoted = [&common_dtype](const Tensor& t) {
+    // dynamic casting on CPU is not allowed
+    return (t.device().is_cpu()) && !(t.scalar_type() == common_dtype)
```

Comment:
If we're only promoting here on CPU, have you thought about what happens on CUDA? It seems that unpromoted self/other tensors would make it to the TensorIteratorConfig.

Reply:
Even TensorIterator does not create temporary of `common_dtype` for GPU (since dynamic casting is supported on CUDA)

https://github.com/pytorch/pytorch/blob/f69cf3cf2ff392d7791b7efbc069a66494492a77/aten/src/ATen/TensorIterator.cpp#L410-L449

cc: @mruberry (to correct if I am wrong in my understanding)